src/cryptography/hazmat/backends/openssl/backend.py

```diff
@@ -1734,6 +1739,114 @@ def _openssh_public_key_bytes(self, key):
                 serialization._ssh_write_string(public_numbers.encode_point())
             )
 
+    def generate_dh_parameters(self, generator, key_size):
+        if key_size < 512:
+            raise ValueError("DH key_size must be at least 512 bits")
+
+        dh_param_cdata = self._lib.DH_new()
+        self.openssl_assert(dh_param_cdata != self._ffi.NULL)
+        dh_param_cdata = self._ffi.gc(dh_param_cdata, self._lib.DH_free)
+
+        res = self._lib.DH_generate_parameters_ex(
+            dh_param_cdata,
+            key_size,
+            generator,
+            self._ffi.NULL
+        )
+        self.openssl_assert(res == 1)
+
+        return _DHParameters(self, dh_param_cdata)
+
+    def generate_dh_private_key(self, parameters):
+        dh_key_cdata = self._lib.DHparams_dup(parameters._dh_cdata)
+        self.openssl_assert(dh_key_cdata != self._ffi.NULL)
+        dh_key_cdata = self._ffi.gc(dh_key_cdata, self._lib.DH_free)
+
+        res = self._lib.DH_generate_key(dh_key_cdata)
+        self.openssl_assert(res == 1)
+
+        return _DHPrivateKey(self, dh_key_cdata)
+
+    def generate_dh_private_key_and_parameters(self, generator, key_size):
+        return self.generate_dh_private_key(
+            self.generate_dh_parameters(generator, key_size))
+
+    def load_dh_private_numbers(self, numbers):
+        parameter_numbers = numbers.public_numbers.parameter_numbers
+
+        dh_cdata = self._lib.DH_new()
+        self.openssl_assert(dh_cdata != self._ffi.NULL)
+        dh_cdata = self._ffi.gc(dh_cdata, self._lib.DH_free)
+
+        p = self._int_to_bn(parameter_numbers.p)
+        g = self._int_to_bn(parameter_numbers.g)
+        pub_key = self._int_to_bn(numbers.public_numbers.y)
+        priv_key = self._int_to_bn(numbers.x)
+
+        res = self._lib.DH_set0_pqg(dh_cdata, p, self._ffi.NULL, g)
+        self.openssl_assert(res == 1)
+
+        res = self._lib.DH_set0_key(dh_cdata, pub_key, priv_key)
+        self.openssl_assert(res == 1)
+
+        codes = self._ffi.new("int[]", 1)
+        res = self._lib.DH_check(dh_cdata, codes)
```

Comment:
This will fail with `DH_UNABLE_TO_CHECK_GENERATOR` for any generator that's not 2 or 5. However, we don't limit to 2/5 in the parameter creation, which means you can write code that fails this check like so:

```python
params = dh.generate_parameters(7, 512, backend)
params.generate_private_key().private_numbers().private_key(backend)
```

To fix this let's limit the generator to 2 or 5 for now (we can revisit whether we need to change how we use `DH_check` at a later date if users show up asking for more generator flexibility)